contrib/mesos/pkg/node/registrator.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"reflect"
+
+	log "github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+const (
+	defaultInboxSize = 1000
+)
+
+type registration struct {
+	hostName string
+	labels   map[string]string
+}
+
+type Registrator struct {
+	store  cache.Store
+	client *client.Client
+	inbox  chan registration
+}
+
+func NewRegistrator(client *client.Client, store cache.Store) *Registrator {
+	return &Registrator{
+		store:  store,
+		client: client,
+		inbox:  make(chan registration, defaultInboxSize),
+	}
+}
+
+func (r *Registrator) Run(terminate <-chan struct{}) error {
+	go func() {
+		for {
+			select {
+			case <-terminate:
+				return
+			case rg := <-r.inbox:
+				n, upToDate, err := r.isUpToDate(rg.hostName, rg.labels)
```

Comment:
before checking this, check for a tie, `terminate` may be closed here and we don't want to process anything in that case